### PR TITLE
feat: show proof size on download button

### PR DESCRIPTION
**Pull request by wackerow**
## Description
- Migration to add column `size` (bigint) to `proofs` table
- Updates `proved/route.ts` to calculate hex byte size and attach to proof `upsert`
- Updates types
- Format value with [pretty-bytes](https://www.npmjs.com/package/pretty-bytes)
- Display on `DownloadButton` at all screen sizes
- Update `seed.ts` to seed random proof size values between 32 KiB and 8 MiB for testing

### CODE REVIEW

**Review by minimalsm — COMMENTED** [review]
hey @wackerow, just a quick sanity check on this as i know it's still in draft.

my understanding is that this calculates the binary size on the client, which might be problematic since it recreates the binary in memory for each rendered button (effectively downloading each file to calculate its size).

as you mentioned, i think it might be better to either add a `proof_size` column to the `proofs` table or calculate the size on the database when fetching proofs data. something like:

```
SELECT 
  proof_id, 
  pg_column_size(proof) / 1024.0 AS proof_size_kb
FROM 
  proofs
ORDER BY 
  proof_size_kb DESC;
```

cc: @pettinarip

File: components/DownloadButton.tsx [hunk 1]
```diff
@@ -26,21 +28,32 @@ const DownloadButton = ({ className, proof }: DownloadButtonProps) => {
...
     "bg-body-secondary/10 hover:bg-body-secondary/10 cursor-auto"
 
-  if (proof_status === "proved")
+  if (proof_status === "proved") {
+    const size = binary
+      ? prettyBytes(
+          new Blob([binary], {
+            type: "application/octet-stream",
+          }).size
+        )
```
[the remark this answers is not in the document]
Comment (wackerow) [reply components/DownloadButton.tsx:37]: yes, draft is for real on this PR... was just trying this out but I think a backend approach is really what we want here

[the remark this answers is not in the document]
Comment (wackerow) [reply components/DownloadButton.tsx]: I wasn't unable to convert this into a selector that could be used with the supabase js client, and we can't use SQL directly.

I had played with adding a column to the proofs table with a trigger function to calculate and add this, but can probably  just get away with a view here, building off what you provided. Downside here is iterating over every proof on the server when the page builds to form the mapping. 

If we add a column to the proofs table, this data would be calculated at time of row creation and would then be available. cc: @pettinarip Curious your take here.

[the remark this answers is not in the document]
Comment (wackerow) [reply components/DownloadButton.tsx]: Pushed the code changes to use a new view `proof_sizes` with `proof_id` and `byte_size`, but doesn't seem to be showing up on Netlify... I cleared cache but haven't pushes any migrations.

[the remark this answers is not in the document]
Comment (wackerow) [reply components/DownloadButton.tsx]: Okay.. chatted with @pettinarip and agree that the simplest approach is to just add a column to the proofs table and we'll calculate the sizes as they proofs are POSTed. Branch updated.

File: components/DownloadButton.tsx [hunk 2]
```diff
@@ -11,35 +13,43 @@ import StatusIcon from "./StatusIcon"
...
+          variant="outline"
+          className={cn(sizingClassName, className)}
+          disabled={!binary}
+          asChild
+        >
+          <a href={`/api/v0/proofs/download/${proof_id}`} download>
+            <ArrowDown />
+            <span className={labelStyle}>
+              <span className={labelDisplay}>
+                Download {size ? "" : "proof"}
```
Comment (minimalsm) [inline components/DownloadButton.tsx:46]: @nloureiro opinion here?

In current implementation, the download label for proofs of 5kb and 10mb would read:

`Download 5kB proof`
`Download 10MB proof`

I think having the size outside of the main text is the most normal pattern to use here.

`Download proof (5kB)`
`Download proof (10MB)`

File: supabase/migrations/20241216091100_add_size_to_proofs.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+alter table public.proofs add column size bigint;
```
Comment (minimalsm) [inline supabase/migrations/20241216091100_add_size_to_proofs.sql:1]: We should specify a unit here e.g. size_bytes

File: components/DownloadButton.tsx (hunk 2 above)
Comment (nloureiro) [reply components/DownloadButton.tsx:46]: yes, this is better
```
Download proof (5kB)
Download proof (10MB)
```

File: app/block/[block]/page.tsx [hunk 4]
```diff
@@ -95,12 +95,12 @@ export default async function BlockDetailsPage({
...
 
-  const { timestamp, block_number, gas_used, proofs, hash } = block
+  const { timestamp, block_number, gas_used, proofs: blockProofs, hash } = block
 
-  const proofsWithTeams = proofs.map((proof) => {
+  const proofs = blockProofs.map((proof) => {
     const team = teams.find((t) => t.user_id === proof.user_id)
     return { ...proof, team }
-  })
+  }) as Proof[]
```
Comment (pettinarip) [inline app/block/[block]/page.tsx:103]: Why are we overriding the inferred type?

**Review by pettinarip — COMMENTED** [review]
Looks good!

fyi I've reset the dev db with an updated seed to see sizes.

File: components/DownloadButton.tsx (hunk 2 above)
Comment (wackerow) [reply components/DownloadButton.tsx:46]: Ok. All buttons will get bigger if we do this approach

File: supabase/migrations/20241216091100_add_size_to_proofs.sql (hunk 3 above)
Comment (wackerow) [reply supabase/migrations/20241216091100_add_size_to_proofs.sql:1]: we don't do this with other units (ie time)

File: app/block/[block]/page.tsx (hunk 4 above)
Comment (wackerow) [reply app/block/[block]/page.tsx:103]: Oops thanks, good catch... It was the new `size` type which needed to be `size?: number | null`

File: components/DownloadButton.tsx (hunk 2 above)
Comment (wackerow) [reply components/DownloadButton.tsx:46]: ...at least that was my thought process in this. These are laid in a column, so there is a shared width. Hence trying to _replace_ "proof" with the size to keep about the same width, but I'm happy to try and squeeze it all in.

Comment (wackerow) [reply components/DownloadButton.tsx:46]: Could also consider two rows:

image

Just ideating cc: @nloureiro

Comment (wackerow) [reply components/DownloadButton.tsx:46]: image

This is the desktop appearance with everything on one line (above suggestion)

...and initial PR suggestion for reference:

image

File: supabase/migrations/20241216091100_add_size_to_proofs.sql (hunk 3 above)
Comment (minimalsm) [reply supabase/migrations/20241216091100_add_size_to_proofs.sql:1]: > _we don't do this with other units (ie time)_

We probably should?

File: app/api/v0/proofs/proved/route.ts [hunk 5]
```diff
@@ -157,6 +157,7 @@ export const POST = withAuth(async ({ request, client, user, timestamp }) => {
         program_id: programId,
         proof_status: "proved",
         proved_timestamp: timestamp,
+        size: Buffer.byteLength(proofHex, "hex"),
```
Comment (pettinarip) [inline app/api/v0/proofs/proved/route.ts:160]:
```suggestion
        size_bytes: Buffer.byteLength(proofHex, "hex"),
```

**Review by pettinarip — APPROVED** [review]

**Pull request merged**